### PR TITLE
Add a viewAsPairs option

### PR DESCRIPTION
**Pull request by danvk**
See #254 
See #313

Syntax for passing options looks like this:

```javascript
  {
    viz: pileup.viz.pileup({
      viewAsPairs: true
    }),
    data: bamSource,
    cssClass: 'tumor',
    name: 'Alignments'
  }
```

I also took the opportunity to rename `playground-complete.html`. The name doesn't make much sense, since there's no other playground that's incomplete!

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/pileup.js/333)

### CODE REVIEW

**Comment by danvk** [comment]
cc @ihodes would appreciate a look-over from you as well, since this is sort of a new user-facing API.

File: src/main/PileupCache.js [hunk 1]
```diff
@@ -49,17 +43,29 @@ class PileupCache {
...
     this.refToPileup = {};
     this.referenceSource = referenceSource;
+    this.viewAsPairs = viewAsPairs;
+  }
+
+  // read name would make a good key, but paired reads from different contigs
+  // shouldn't be grouped visually. Hence we use read name + contig.
+  groupKey(read: Alignment): string {
+    if (this.viewAsPairs) {
+      return `${read.name}:${read.ref}`;
```
Comment (ihodes) [inline src/main/PileupCache.js]: is ref == contig?

File: src/main/PileupCache.js [hunk 2]
```diff
@@ -91,13 +97,17 @@ class PileupCache {
...
       // TODO: if the mate Alignment is also available, it would be better to use that.
-      var intervals = [range];
-      var mateProps = read.getMateProperties();
-      if (mateProps && mateProps.ref && mateProps.ref == read.ref) {
-        mateInterval = new ContigInterval(mateProps.ref, mateProps.pos, mateProps.pos + range.length());
-        intervals.push(mateInterval);
+      if (this.viewAsPairs) {
+        var mateProps = read.getMateProperties();
+        var intervals = [range];
+        if (mateProps && mateProps.ref && mateProps.ref == read.ref) {
```
Comment (ihodes) [inline src/main/PileupCache.js]: Not at all necessary, just a thought; I've enjoyed using functions like `equalsIn(obj, path, val)` to handle this (in this case, `equalsIn(matProps, ['ref'], read.ref)`).

File: src/main/PileupCache.js [hunk 3]
```diff
@@ -160,7 +170,7 @@ function spanAndInsert(intervals: ContigInterval<string>[]) {
   if (intervals.length == 1) {
     return {insert: null, span: intervals[0]};
   } else if (intervals.length !=2) {
-    throw `Called spanAndInsert with ${intervals.length} \in [1, 2]`;
+    throw `Called spanAndInsert with ${intervals.length} \notin [1, 2]`;
```
Comment (ihodes) [inline src/main/PileupCache.js]: Don't you wish the console would render LaTeX?

**Comment by armish** [comment]
one speculative comment to get a conversation going about alternative ways to handle options, but this looks great!


---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/pileup.js, line 135 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/333#-K1a42AYDvKXaMr24atP)** ([raw file](https://github.com/hammerlab/pileup.js/blob/b4247631cee9040c16da707b6944b826053e6b6d/src/main/pileup.js#L135)):</sup>
when I proposed #313, I was thinking about embedding default options to the tracks and make pileup pass an `opts` object to all of the components by default (whether empty or not); e.g.:

```
{
    viz: pileup.viz.pileup(),
    data: bamSource,
    cssClass: 'tumor',
    name: 'Alignments',
    opts: {viewAsPairs: true}
  },
{
    viz: pileup.viz.coverage(),
    data: bamSource,
    cssClass: 'tumor',
    name: 'Coverage',
    opts: {showMismatches: true}
  }
```
where tracks are responsible for handling whatever options are passed to them (e.g. overriding default ones with the incoming ones)

Looks like yours is a slightly different approach where instead of tracks, it is the pileup that knows about the default options, if I understand this correctly. I like this as it handles options in a generic way across tracks and doesn't lead to repetitive `_.extend` calls within tracks; but I don't like the fact that default options are hidden somewhere else in the code, which might make it harder to read the code for individual tracks and understand what each option does.

I don't really have strong opinions against either way, but love to hear what you guys think about the alternative?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/333)

**Comment by ihodes** [comment]
@danvk LGTM, I like the option-passing (taking advantage of that decision early on, now).

**Comment by danvk** [comment]
I've addressed all comments. Let me know what you guys think.


---

Review status: 9 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[src/main/pileup.js, line 135 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/333#-K1a42AYDvKXaMr24atP)** ([raw file](https://github.com/hammerlab/pileup.js/blob/b4247631cee9040c16da707b6944b826053e6b6d/src/main/pileup.js#L135)):</sup>
You're totally right that tracks should set their own default options. I've added a `defaultOptions` property to `PileupTrack`.

I think it does make sense to pass option separately to the data source and the visualization (there will naturally be different options which are relevant to each). How to get two pileup tracks to share visualization options is an interesting question.


---

<sup>**[src/main/PileupCache.js, line 59 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/333#-K1_zq3bRSSXNuEiANtp-r1-59)** ([raw file](https://github.com/hammerlab/pileup.js/blob/b4247631cee9040c16da707b6944b826053e6b6d/src/main/PileupCache.js#L59)):</sup>
Yep, this is what Samtools calls it.


---

<sup>**[src/main/PileupCache.js, line 103 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/333#-K1_zq3bRSSXNuEiANtp-r1-103)** ([raw file](https://github.com/hammerlab/pileup.js/blob/b4247631cee9040c16da707b6944b826053e6b6d/src/main/PileupCache.js#L103)):</sup>
Maybe I'm too used to static languages, but `mateProps.ref` and `mateProps['ref']` feel like different things to me. (I realize they're exactly the same in JS).

Actually, there is a reason to prefer writing it this way. Flow can analyze it one way but not the other (e.g. it would catch if I wrote `contig` instead of `ref`).


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/333)

**Comment by armish** [comment]
Awesome! :shipit: 


---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/333)

**Pull request merged**